text/0012-jsonfile.md

```diff
@@ -0,0 +1,60 @@
+- Frequent writes will cause the entire file to be written to disk repeatedly, [which causes complaints about SSD impact](https://bugzilla.mozilla.org/show_bug.cgi?id=1304389).
+- The entire file typically needs to be read into memory to be used, which increases memory footprint and can impact startup time.
+- Writes at shutdown harm the user experience and don't happen during a crash.
+- In order to achieve atomic file writes, the entire file contents briefly exist twice on disk, which could be problematic on mobile platforms for large data sets.
+- The only approach to cross-process use is duplication of all or part of the in-memory data, which can increase memory footprint. It's not possible for multiple processes to collaborate on the same data via the filesystem. This is why Firefox for Android still uses the SQLite implementation of `nsILoginManagerStorage`: it allows the Android-side `ContentProvider` code to read and write saved logins without worrying about Gecko, which nominally owns the backing storage, but has a shorter lifetime than the enclosing Android code. Using desktop's JSON-backed storage on Android would require launching Gecko on each sync, using messaging for safe access to the file contents.
+- The in-memory object is essentially a non-transactional write-back cache. This has several issues:
+  - It makes isolation (readers can't see in-progress writes) and atomicity (all writes complete or none do) difficult: we typically mutate the in-memory object directly, and so an exception can cause partial changes to 'commit', and readers will see each change as it is applied.
+  - Similiarly, timed flushing makes consistency difficult: it's possible for only some of a set of asynchronous writes to be flushed to disk because the flush beat the last few writes, leaving the data inconsistent after a crash.
+    
+    JSONFile.jsm advises that callers do all of their work synchronously on the main thread to avoid the possibility of concurrent readers or partial writes:
+    
+    > The raw data should be manipulated synchronously, without waiting for the
+    > event loop or for promise resolution, so that the saved file is always
+    > consistent. 
+     
+    This can cause jank.
+- Synchronization of data stored in this way takes some careful thought. Syncs typically take hundreds of milliseconds or more, and involve asynchronous network operations, which makes exclusive synchronous access to the in-memory data between reads and writes infeasible.
+- There is a tension between durability (that is: writes that complete are permanent) and performance. We typically choose not to flush the file immediately and synchronously after every change, but not doing so leaves a window in which a crash would cause data loss. By default, that window is 1.5 seconds, plus the time needed to write and atomically switch the files. This forces careful consumers to manually flush if they want their writes to stick, which is a bad pattern.
```

Comment:
The alternative would be always flushing which would be worse for performance so I think it makes sense to give the consumer the choice.

Reply:
I agree that a choice is better than no choice in JSONFile's case.

The main alternative, IMO, is to not routinely force consumers to trade performance versus durability — stores exist that don't require an entire file to be written for a single small change.